ambry-router/src/test/java/com/github/ambry/router/InMemoryCloudDestinationErrorSimulationTest.java

```diff
@@ -0,0 +1,387 @@
+/**
+ * Copyright 2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.router;
+
+
+import com.github.ambry.account.InMemAccountService;
+import com.github.ambry.cloud.CloudDestination;
+import com.github.ambry.cloud.CloudDestinationFactory;
+import com.github.ambry.cloud.LatchBasedInMemoryCloudDestinationFactory;
+import com.github.ambry.clustermap.MockClusterMap;
+import com.github.ambry.clustermap.PartitionId;
+import com.github.ambry.clustermap.ReplicaId;
+import com.github.ambry.clustermap.ReplicaType;
+import com.github.ambry.commons.BlobId;
+import com.github.ambry.commons.LoggingNotificationSystem;
+import com.github.ambry.config.CloudConfig;
+import com.github.ambry.config.NetworkConfig;
+import com.github.ambry.config.RouterConfig;
+import com.github.ambry.config.VerifiableProperties;
+import com.github.ambry.messageformat.BlobProperties;
+import com.github.ambry.messageformat.BlobType;
+import com.github.ambry.network.CompositeNetworkClientFactory;
+import com.github.ambry.network.LocalNetworkClientFactory;
+import com.github.ambry.network.LocalRequestResponseChannel;
+import com.github.ambry.network.NetworkClient;
+import com.github.ambry.network.NetworkClientFactory;
+import com.github.ambry.network.NetworkMetrics;
+import com.github.ambry.network.Port;
+import com.github.ambry.network.PortType;
+import com.github.ambry.network.RequestInfo;
+import com.github.ambry.network.ResponseInfo;
+import com.github.ambry.protocol.PutRequest;
+import com.github.ambry.protocol.RequestHandlerPool;
+import com.github.ambry.server.ServerErrorCode;
+import com.github.ambry.store.StoreErrorCodes;
+import com.github.ambry.utils.MockTime;
+import com.github.ambry.utils.NettyByteBufLeakHelper;
+import com.github.ambry.utils.Utils;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.PooledByteBufAllocator;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.EnumMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.Properties;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.junit.Assert.*;
+
+public class InMemoryCloudDestinationErrorSimulationTest {
```

Comment:
I think you are making these cases too complicated. This unit test should only test is the LatchBasedInMemoryCloudDestination would be able to return various errors as we design. It's nice to have it wrapped inside a cloud storage manager and a request handler. But having routers to send the request is too much for just testing the error simulation. 

I would just remove the router code and only rely on the request handler. And we can send the PutRequest, GetRequest, etc to the request handler and assert that we get errors we want back from the request handler.

Reply:
Thanks for the suggestion. The next revision use snetworkClient.sendAndPoll.